File: cache/src/main/java/com/osrsd/cache/provider/SequenceProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.Definition;
import com.osrsd.cache.def.SequenceDefinition;
import com.osrsd.cache.util.ByteBufferExt;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;

@Slf4j
public final class SequenceProvider {

    public static Definition decode(ByteBuffer buffer, SequenceDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            int length;
            int index;
            if (opcode == 1) {
                length = buffer.getShort() & 0xffff;
                definition.setFrameLengths(new int[length]);
                for (index = 0; index < length; ++index) {
                    definition.getFrameLengths()[index] = buffer.getShort() & 0xffff;
                }
                definition.setFrameIDs(new int[length]);
                for (index = 0; index < length; ++index) {
                    definition.getFrameIDs()[index] = buffer.getShort() & 0xffff;
                }
                for (index = 0; index < length; ++index) {
                    definition.getFrameIDs()[index] += (buffer.getShort() & 0xffff) << 16;
                }
            } else if (opcode == 2) {
                definition.setFrameStep(buffer.getShort() & 0xffff);
            } else if (opcode == 3) {
                length = buffer.get() & 0xff;
                definition.setInterleaveLeave(new int[1 + length]);
                for (index = 0; index < length; ++index) {
                    definition.getInterleaveLeave()[index] = buffer.get() & 0xff;
                }
                definition.getInterleaveLeave()[length] = 9999999;
            } else if (opcode == 4) {
                definition.setStretches(true);
            } else if (opcode == 5) {
                definition.setForcedPriority(buffer.get() & 0xff);
            } else if (opcode == 6) {
                definition.setLeftHandItem(buffer.getShort() & 0xffff);
            } else if (opcode == 7) {
                definition.setRightHandItem(buffer.getShort() & 0xffff);
            } else if (opcode == 8) {
                definition.setMaxLoops(buffer.get() & 0xff);
            } else if (opcode == 9) {
                definition.setPrecedenceAnimating(buffer.get() & 0xff);
            } else if (opcode == 10) {
                definition.setPriority(buffer.get() & 0xff);
            } else if (opcode == 11) {
                definition.setReplyMode(buffer.get() & 0xff);
            } else if (opcode == 12) {
                length = buffer.get() & 0xff;
                definition.setChatFrameIds(new int[length]);
                for (index = 0; index < length; ++index) {
                    definition.getChatFrameIds()[index] = buffer.getShort() & 0xffff;
                }
                for (index = 0; index < length; ++index) {
                    definition.getChatFrameIds()[index] += (buffer.getShort() & 0xffff) << 16;
                }
            } else if (opcode == 13) {
                length = buffer.get() & 0xff;
                definition.setFrameSounds(new int[length]);
                for (index = 0; index < length; ++index) {
                    definition.getFrameSounds()[index] = ByteBufferExt.getMedium(buffer);
                }
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/SequenceProvider.kt
package com.osedit.cache.provider

import com.osedit.cache.util.ByteBufferExt
import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.SequenceDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object SequenceProvider : Provider<SequenceDefinition> {

    private val log: Logger = LoggerFactory.getLogger(SequenceProvider::class.java)

    override fun decode(buffer: ByteBuffer, definition: SequenceDefinition): Definition {
        do {
            when (val opcode: Int = buffer.get().toInt() and 0xff) {
                1 -> {
                    val length: Int = buffer.short.toInt() and 0xffff
                    definition.frameLengths = IntArray(length)
                    (0 until length).forEach {
                        definition.frameLengths!![it] = buffer.short.toInt() and 0xffff
                    }
                    definition.frameIDs = IntArray(length)
                    (0 until length).forEach {
                        definition.frameIDs!![it] = buffer.short.toInt() and 0xffff
                    }
                    (0 until length).forEach {
                        definition.frameIDs!![it] += (buffer.short.toInt() and 0xffff) shl 16
                    }
                }
                2 -> definition.frameStep = buffer.short.toInt() and 0xffff
                3 -> {
                    val length: Int = buffer.get().toInt() and 0xff
                    definition.interleaveLeave = IntArray(length + 1)
                    (0 until length).forEach {
                        definition.interleaveLeave!![it] = buffer.get().toInt() and 0xff
                    }
                    definition.interleaveLeave!![length] = 9_999_999
                }
                4 -> definition.stretches = true
                5 -> definition.forcedPriority = buffer.get().toInt() and 0xff
                6 -> definition.leftHandItem = buffer.short.toInt() and 0xffff
                7 -> definition.rightHandItem = buffer.short.toInt() and 0xffff
                8 -> definition.maxLoops = buffer.get().toInt() and 0xff
                9 -> definition.precedenceAnimating = buffer.get().toInt() and 0xff
                10 -> definition.priority = buffer.get().toInt() and 0xff
                11 -> definition.replyMode = buffer.get().toInt() and 0xff
                12 -> {
                    val length: Int = buffer.get().toInt() and 0xff
                    definition.chatFrameIds = IntArray(length)
                    (0 until length).forEach {
                        definition.chatFrameIds!![it] = buffer.short.toInt() and 0xffff
                    }
                    (0 until length).forEach {
                        definition.chatFrameIds!![it] += (buffer.short.toInt() and 0xffff) shl 16
                    }
                }
                13 -> {
                    val length: Int = buffer.get().toInt() and 0xff
                    definition.frameSounds = IntArray(length)
                    (0 until length).forEach {
                        definition.frameSounds!![it] = ByteBufferExt.getMedium(buffer)
                    }
                }
                0 -> break
                else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
            }
        } while (true)
        return definition
    }

}
